File: Components/DevExpressVCL/Demos/ExpressScheduler/CBuilder/ResourceDemo/RentUnit.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "..\cxDemosBCB.inc"
#include "RentUnit.h"
#include "ResourceMainUnit.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxDateNavigator"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxListBox"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxScheduler"
#pragma link "cxSchedulerCustomControls"
#pragma link "cxSchedulerCustomResourceView"
#pragma link "cxSchedulerDateNavigator"
#pragma link "cxSchedulerDayView"
#pragma link "cxSchedulerStorage"
#pragma link "cxStyles"
#pragma link "cxTextEdit"
#pragma resource "*.dfm"

int __fastcall CompareEvents(TcxSchedulerEvent *AEvent1, TcxSchedulerEvent *AEvent2)
{
  if (AEvent1->Start == AEvent2->Start)
    return 0;
  else
    if (AEvent1->Start < AEvent2->Start) 
      return -1;
    else
      return 1;
}

//---------------------------------------------------------------------------
__fastcall TfrmRentCar::TfrmRentCar(TComponent* Owner)
        : TForm(Owner)
{
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::AddPeriodTime(TDateTime AStart, TDateTime AFinish)
{
  if (AStart == AFinish) return;
  FPeriods->Add(TimeToDateTime(AStart));
  FPeriods->Add(TimeToDateTime(AFinish));
  lbxPeriod->Items->Add(TimeToStr(AStart)  + '-' + TimeToStr(AFinish));
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::AddRent(TDateTime AStart, TDateTime AFinish)
{
  if (AStart >= AFinish) return;
  TcxSchedulerEvent *AEvent = Storage()->createEvent();
  AEvent->Start = AStart;
  AEvent->Finish = AFinish;
  AEvent->Caption = edtUserName->Text;
  AEvent->ResourceID = GetIndex(lvCars);
}
//---------------------------------------------------------------------------

TDateTime __fastcall TfrmRentCar::DateTimeToTime(TDateTime ATime)
{
  return DateNavigator->Date + EncodeTime((double)ATime / 60, (int)ATime % 60, 0, 0);
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::FillTimeList()
{
  TDateTime AFinish;
  lbxPeriod->Items->BeginUpdate();
  FPeriods->Clear();
  try {
    lbxPeriod->Items->Clear();
    TDateTime AStart = EncodeTime(0, 0, 0, 0);
    if (FDayEvents->Count == 0)
      AddPeriodTime(AStart, EncodeTime(23, 59, 0, 0));
    else
      for (int I = 0; I <= FDayEvents->Count; I++) {
        if (I > 0)
          AStart = FDayEvents->Items[I - 1]->Finish;
        if (I < FDayEvents->Count)
          AFinish = FDayEvents->Items[I]->Start;
        else
          AFinish = EncodeTime(23, 59, 0, 0);
        AddPeriodTime(AStart, AFinish);
      }
  }
  __finally {
    lbxPeriod->Items->EndUpdate();
  }
}

//---------------------------------------------------------------------------
bool __fastcall TfrmRentCar::Intersect(TDateTime AStart1, TDateTime AFinish1,
  TDateTime AStart2, TDateTime AFinish2)
{
  return (AStart1 < AFinish2) && (AStart2 < AFinish1);
}

//---------------------------------------------------------------------------
bool __fastcall TfrmRentCar::IntersectTime(TDateTime &AStart1,
  TDateTime &AFinish1, TDateTime &AStart2, TDateTime &AFinish2)
{
  if (AStart2 > AStart1)
    AStart1 = AStart2;
  if (AFinish2 < AFinish1)
    AFinish1 = AFinish2;
  return AFinish1 > AStart1;
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::ProcessSelectItem(int AIndex)
{
  FDays->Clear();
  Storage()->GetEvents(FEventsList, DateNavigator->RealFirstDate,
    DateNavigator->RealLastDate, AIndex);
  for (int I = (int)DateNavigator->RealFirstDate; I <= (int)DateNavigator->RealLastDate; I++)
    for (int J = 0; J < FEventsList->Count; J++)
      if (FEventsList->Items[J]->IsDayEvent(I)) {
        FDays->Add(I);
        break;
      }
  DateNavigator->LayoutChanged();
  DateNavigatorSelectionChanged(DateNavigator, DateNavigator->Date, DateNavigator->Date);
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::RentCar()
{
  TDateTime AStart1, AFinish1;
  TDateTime AStart = TimeScheduler->SelStart;
  TDateTime AFinish = TimeScheduler->SelFinish;
  try {
    for (int I = 0; I <= FDayEvents->Count; I++) {
      if (I == 0)
        AStart1 = (int)AStart;
      if (I == FDayEvents->Count)
        AFinish1 = (int)AStart + (double)EncodeTime(23, 59, 0, 0);
      else
        AFinish1 = FDayEvents->Items[I]->Start;
      if ((AStart1 < AFinish1) && IntersectTime(AStart1, AFinish1, AStart, AFinish))
        AddRent(AStart1, AFinish1);
      if ((FDayEvents->Count > 0) && (I < FDayEvents->Count))
        AStart1 = FDayEvents->Items[I]->Finish;
    }
  }
  __finally {
    Storage()->PostEvents();
    ProcessSelectItem(GetIndex(lvCars));
    TimeScheduler->Refresh();
  }
}

//---------------------------------------------------------------------------
TcxSchedulerStorage* __fastcall TfrmRentCar::Storage()
{
  return ResourceDemoMainForm->Storage;
}

//---------------------------------------------------------------------------
TDateTime __fastcall TfrmRentCar::TimeToDateTime(TDateTime ATime)
{
  WORD H, M, S, MS;
  DecodeTime(ATime, H, M, S, MS);
  return H * 60 + M;
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::FormCreate(TObject* Sender)
{
  FPeriods = new TcxSchedulerDateList;
  FEventsList = new TcxSchedulerFilteredEventList;
  FDays = new TcxSchedulerDateList;
  FDayEvents = new TcxSchedulerEventList;
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::FormDestroy(TObject* Sender)
{
  delete FPeriods;
  delete FEventsList;
  delete FDays;
  delete FDayEvents;
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::FormShow(TObject* Sender)
{
  Storage()->FullRefresh();
  TcxCustomScheduler* AScheduler = ResourceDemoMainForm->Scheduler;
  DateNavigator->Date = (int)AScheduler->SelStart;
  if (AScheduler->SelResource == NULL)
    SetIndex(lvCars, 0);
  else
    SetIndex(lvCars, AScheduler->SelResource->ID);
  ProcessSelectItem(GetIndex(lvCars));
  TimeScheduler->SelectDays(DateNavigator->Date, DateNavigator->Date, true);
  TimeScheduler->SelectTime(AScheduler->SelStart, AScheduler->SelFinish, NULL);
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::TimeSchedulerCustomDrawContent(TObject* Sender,
  TcxCanvas* ACanvas, TcxSchedulerContentCellViewInfo* AViewInfo, bool &ADone)
{
  for (int I = 0; I < FDayEvents->Count; I++)
    if (Intersect(AViewInfo->TimeStart, AViewInfo->TimeFinish,
      FDayEvents->Items[I]->Start, FDayEvents->Items[I]->Finish)) {
      ACanvas->Brush->Color = clBtnShadow;
      ACanvas->FillRect(AViewInfo->Bounds, NULL, false);
      ACanvas->Font->Color = clBtnText;
      ACanvas->DrawTexT("Busy time", AViewInfo->Bounds, cxAlignCenter, true);
      AViewInfo->Transparent = true;
    }
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::DateNavigatorCustomDrawDayNumber(TObject* Sender,
  TcxCanvas* ACanvas, TcxSchedulerDateNavigatorDayNumberViewInfo* AViewInfo,
  bool &ADone)
{
  if (FDays->IndexOf(AViewInfo->Date) != -1)
    ACanvas->Font = cxBoldStyle->Font;
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::DateNavigatorCustomDrawContent(TObject* Sender,
  TcxCanvas* ACanvas, TcxSchedulerDateNavigatorMonthContentViewInfo* AViewInfo,
  bool &ADone)
{
  ResourceDemoMainForm->DrawDateNavigatorContent(ACanvas, AViewInfo, ADone);
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::lvCarsSelectItem(TObject* Sender, TListItem* Item,
  bool Selected)
{
  if (!Selected) return;
  ProcessSelectItem(Item->Index);
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::lbxPeriodClick(TObject* Sender)
{
  int AIndex = lbxPeriod->ItemIndex;
  if (AIndex != -1) {
    TimeScheduler->SelectTime(DateTimeToTime(FPeriods->Items[AIndex * 2]),
      DateTimeToTime((double)FPeriods->Items[AIndex * 2 + 1]) + (double)HourToTime, NULL);
  }
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::TimeSchedulerKeyDown(TObject* Sender, WORD &Key,
  TShiftState Shift)
{
  if (Key != VK_TAB) return;
  TWinControl* AControl = FindNextControl(TimeScheduler,
    !Shift.Contains(ssCtrl), true, true);
  if (AControl != NULL)
    AControl->SetFocus();
}

//---------------------------------------------------------------------------
void __fastcall TfrmRentCar::btnRentClick(TObject* Sender)
{
  RentCar();
}
//---------------------------------------------------------------------------

void __fastcall TfrmRentCar::DateNavigatorSelectionChanged(TObject *Sender,
      const TDateTime AStart, const TDateTime AFinish)
{
  FDayEvents->Clear();
  for (int I = 0; I < FEventsList->Count; I++)
    if (FEventsList->Items[I]->IsDayEvent(AStart))
      FDayEvents->Add(FEventsList->Items[I]);
  FDayEvents->Sort(CompareEvents);
  FillTimeList();
  TimeScheduler->SelectDays(AStart, AFinish, true);
  lbxPeriod->ItemIndex = 0;
  lbxPeriodClick(lbxPeriod);
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressScheduler/Delphi/ResourceDemo/RentUnit.pas
unit RentUnit;

interface

{$I cxVer.inc}

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxStyles, cxGraphics, cxEdit, cxScheduler, cxSchedulerUtils,
  cxSchedulerCustomControls, cxSchedulerCustomResourceView, cxSchedulerStorage,
  cxSchedulerDayView, cxSchedulerDateNavigator, StdCtrls, cxContainer,
  cxDateNavigator, cxControls, ComCtrls, ExtCtrls, Math, cxDateUtils,
  DateUtils,cxLookAndFeelPainters, cxButtons, cxTextEdit, cxListBox, cxListView,
  ToolWin, Menus, cxSchedulerTimeGridView, cxSchedulerWeekView,
  cxSchedulerYearView;

type
  TfrmRentCar = class(TForm)
    lvCars: TListView;
    DateNavigator: TcxDateNavigator;
    edtUserName: TcxTextEdit;
    lbxPeriod: TcxListBox;
    TimeScheduler: TcxScheduler;
    btnRent: TcxButton;
    btnCancel: TcxButton;
    lbChoosePeriod: TLabel;
    lbChooseDate: TLabel;
    lbChooseCar: TLabel;
    lbCustomerName: TLabel;
    lbChooseTime: TLabel;
    cxStyleRepository: TcxStyleRepository;
    cxBoldStyle: TcxStyle;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure TimeSchedulerCustomDrawContent(Sender: TObject;
      ACanvas: TcxCanvas; AViewInfo: TcxSchedulerContentCellViewInfo;
      var ADone: Boolean);
    procedure DateNavigatorSelectionChanged(Sender: TObject;
      const AStart, AFinish: TDateTime);
    procedure DateNavigatorCustomDrawDayNumber(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerDateNavigatorDayNumberViewInfo;
      var ADone: Boolean);
    procedure DateNavigatorCustomDrawContent(Sender: TObject;
      ACanvas: TcxCanvas;
      AViewInfo: TcxSchedulerDateNavigatorMonthContentViewInfo;
      var ADone: Boolean);
    procedure lvCarsSelectItem(Sender: TObject; Item: TListItem;
      Selected: Boolean);
    procedure lbxPeriodClick(Sender: TObject);
    procedure TimeSchedulerKeyDown(Sender: TObject; var Key: Word;
      Shift: TShiftState);
    procedure btnRentClick(Sender: TObject);
  private
    FDayEvents: TcxSchedulerEventList;
    FDays: TcxSchedulerDateList;
    FEventsList: TcxSchedulerFilteredEventList;
    FPeriods: TcxSchedulerDateList;
    procedure AddPeriodTime(AStart, AFinish: TDateTime);
    procedure AddRent(AStart, AFinish: TDateTime);
    function DateTimeToTime(ATime: TDateTime): TDateTime;
    procedure FillTimeList;
    function Intersect(AStart1, AFinish1, AStart2, AFinish2: TDateTime): Boolean;
    function IntersectTime(var AStart1, AFinish1, AStart2, AFinish2: TDateTime): Boolean;
    procedure ProcessSelectItem(AIndex: Integer);
    procedure RentCar;
    function Storage: TcxSchedulerStorage;
    function TimeToDateTime(ATime: TDateTime): TDateTime;
  end;

implementation

uses
  ResourceMainUnit;


{$R *.dfm}

function CompareEvents(AEvent1, AEvent2: TcxSchedulerEvent): Integer;
begin
  if AEvent1.Start = AEvent2.Start then
    Result := 0
  else
    if AEvent1.Start < AEvent2.Start then
      Result := -1
    else
      Result := 1;
end;

procedure TfrmRentCar.AddPeriodTime(AStart, AFinish: TDateTime);
begin
  if AStart = AFinish then Exit;
  FPeriods.Add(TimeToDateTime(AStart));
  FPeriods.Add(TimeToDateTime(AFinish));
  lbxPeriod.Items.Add(TimeToStr(AStart)  + '-' + TimeToStr(AFinish));
end;

procedure TfrmRentCar.AddRent(AStart, AFinish: TDateTime);
begin
  if AStart >= AFinish then Exit;
  with Storage.createEvent do
  begin
    Start := AStart;
    Finish := AFinish;
    Caption := edtUserName.Text;
    ResourceID := GetIndex(lvCars);
  end;
end;

function TfrmRentCar.DateTimeToTime(ATime: TDateTime): TDateTime;
begin
  Result := DateNavigator.Date +
    EncodeTime(Round(ATime) div 60, Round(ATime) mod 60, 0, 0);
end;

procedure TfrmRentCar.FillTimeList;
var
  I: Integer;
  AInfo: TcxSchedulerEventConflictsInfo;
begin
  lbxPeriod.Items.BeginUpdate;
  try
    FPeriods.Clear;
    lbxPeriod.Items.Clear;
    AInfo := TcxSchedulerEventConflictsInfo.Create(
      Storage, False, DateNavigator.Date + EncodeTime(0, 0, 0, 0),
      DateNavigator.Date + EncodeTime(23, 59, 0, 0), GetIndex(lvCars));
    try
      for I := 0 to AInfo.TimeRanges.Count - 1 do
        AddPeriodTime(AInfo.TimeRanges[I].Start, AInfo.TimeRanges[I].Finish);
    finally
      AInfo.Free;
    end;
  finally
    lbxPeriod.Items.EndUpdate;
  end;
end;

function TfrmRentCar.Intersect(AStart1, AFinish1, AStart2, AFinish2: TDateTime): Boolean;
begin
  Result := (AStart1 < AFinish2) and (AStart2 < AFinish1);
end;

function TfrmRentCar.IntersectTime(
  var AStart1, AFinish1, AStart2, AFinish2: TDateTime): Boolean;
begin
  if AStart2 > AStart1 then
    AStart1 := AStart2;
  if AFinish2 < AFinish1 then
    AFinish1 := AFinish2;
  Result := AFinish1 > AStart1;
end;

procedure TfrmRentCar.ProcessSelectItem(AIndex: Integer);
var
  I, J: Integer;
begin
  FDays.Clear;
  Storage.GetEvents(FEventsList, DateNavigator.RealFirstDate,
    DateNavigator.RealLastDate, AIndex);
  for I := Trunc(DateNavigator.RealFirstDate) to Trunc(DateNavigator.RealLastDate) do
    for J := 0 to FEventsList.Count - 1 do
      if FEventsList[J].IsDayEvent(I) then
      begin
        FDays.Add(I);
        Break
      end;
  DateNavigator.LayoutChanged;
  DateNavigatorSelectionChanged(DateNavigator, DateNavigator.Date, DateNavigator.Date);
end;

procedure TfrmRentCar.RentCar;
var
  I: Integer;
  AStart, AFinish, AStart1, AFinish1: TDateTime;
begin
  AStart := TimeScheduler.SelStart;
  AFinish := TimeScheduler.SelFinish;
  try
    for I := 0 to FDayEvents.Count do
    begin
      if I = 0 then
        AStart1 := DateOf(AStart);
      if I = FDayEvents.Count then
        AFinish1 := DateOf(AStart) + EncodeTime(23, 59, 0, 0)
      else
        AFinish1 := FDayEvents[I].Start;
      if (AStart1 < AFinish1) and IntersectTime(AStart1, AFinish1, AStart, AFinish) then
        AddRent(AStart1, AFinish1);
      if (FDayEvents.Count > 0) and (I < FDayEvents.Count) then
        AStart1 := FDayEvents[I].Finish;
    end;
  finally
    Storage.PostEvents;
    ProcessSelectItem(GetIndex(lvCars));
    TimeScheduler.Refresh;
  end;
end;

function TfrmRentCar.Storage: TcxSchedulerStorage;
begin
  Result := ResourceDemoMainForm.Storage;
end;

function TfrmRentCar.TimeToDateTime(ATime: TDateTime): TDateTime;
var
  H, M, S, MS: Word;
begin
  DecodeTime(ATime, H, M, S, MS);
  Result := H * 60 + M;
end;

procedure TfrmRentCar.FormCreate(Sender: TObject);
begin
  FPeriods := TcxSchedulerDateList.Create;
  FEventsList := TcxSchedulerFilteredEventList.Create;
  FDays := TcxSchedulerDateList.Create;
  FDayEvents := TcxSchedulerEventList.Create;
end;

procedure TfrmRentCar.FormDestroy(Sender: TObject);
begin
  FreeAndNil(FEventsList);
  FreeAndNil(FDays);
  FreeAndNil(FDayEvents);
  FreeAndNil(FPeriods);
end;

procedure TfrmRentCar.FormShow(Sender: TObject);
var
  AScheduler: TcxCustomScheduler;
begin
  Storage.FullRefresh;
  AScheduler := ResourceDemoMainForm.Scheduler;
  if AScheduler.SelResource = nil then
    SetIndex(lvCars, 0)
  else
    SetIndex(lvCars, AScheduler.SelResource.ID);
  ProcessSelectItem(GetIndex(lvCars));
  DateNavigator.Date := DateOf(AScheduler.SelStart);
  ProcessSelectItem(GetIndex(lvCars));
  TimeScheduler.SelectDays(DateNavigator.Date, DateNavigator.Date, True);
  TimeScheduler.SelectTime(AScheduler.SelStart, AScheduler.SelFinish, nil);
end;

procedure TfrmRentCar.TimeSchedulerCustomDrawContent(Sender: TObject;
  ACanvas: TcxCanvas; AViewInfo: TcxSchedulerContentCellViewInfo;
  var ADone: Boolean);
var
  I: Integer;
const
  scxBusyTime = 'Busy time';
begin
  for I := 0 to FDayEvents.Count - 1 do
    if Intersect(AViewInfo.TimeStart, AViewInfo.TimeFinish,
      FDayEvents[I].Start, FDayEvents[I].Finish) then
    begin
      ACanvas.Brush.Color := clBtnShadow;
      ACanvas.FillRect(AViewInfo.Bounds);
      ACanvas.Font.Color := clBtnText;
      ACanvas.DrawTexT(scxBusyTime, AViewInfo.Bounds, cxAlignCenter);
      AViewInfo.Transparent := True;
    end;
end;

procedure TfrmRentCar.DateNavigatorSelectionChanged(Sender: TObject;
  const AStart, AFinish: TDateTime);
var
  I: Integer;
begin
  FDayEvents.Clear;
  for I := 0 to FEventsList.Count - 1 do
    if FEventsList[I].IsDayEvent(Trunc(AStart)) then
      FDayEvents.Add(FEventsList[I]);
  FDayEvents.Sort(CompareEvents);
  FillTimeList;
  TimeScheduler.SelectDays(AStart, AFinish, True);
  lbxPeriod.ItemIndex := 0;
  lbxPeriodClick(lbxPeriod);
end;

procedure TfrmRentCar.DateNavigatorCustomDrawDayNumber(Sender: TObject;
  ACanvas: TcxCanvas;
  AViewInfo: TcxSchedulerDateNavigatorDayNumberViewInfo;
  var ADone: Boolean);
begin
  if FDays.IndexOf(AViewInfo.Date) <> -1 then
    ACanvas.Font := cxBoldStyle.Font;
end;

procedure TfrmRentCar.DateNavigatorCustomDrawContent(Sender: TObject;
  ACanvas: TcxCanvas; AViewInfo: TcxSchedulerDateNavigatorMonthContentViewInfo;
  var ADone: Boolean);
begin
  ResourceDemoMainForm.DrawDateNavigatorContent(ACanvas, AViewInfo, ADone);
end;

procedure TfrmRentCar.lvCarsSelectItem(Sender: TObject; Item: TListItem;
  Selected: Boolean);
begin
  if not Selected then Exit;
  ProcessSelectItem(Item.Index);
end;

procedure TfrmRentCar.lbxPeriodClick(Sender: TObject);
var
  AIndex: Integer;
begin
  AIndex := lbxPeriod.ItemIndex;
  if AIndex <> -1 then
  begin
    TimeScheduler.SelectTime(DateTimeToTime(FPeriods[AIndex * 2]),
      DateTimeToTime(FPeriods[AIndex * 2 + 1]) + HourToTime, nil);
  end;
end;

procedure TfrmRentCar.TimeSchedulerKeyDown(Sender: TObject; var Key: Word;
  Shift: TShiftState);
var
  AControl: TWinControl;
begin
  if Key <> VK_TAB then Exit;
  AControl := FindNextControl(TimeScheduler, not (ssCtrl in Shift), True, True);
  if AControl <> nil then
    AControl.SetFocus;
end;

procedure TfrmRentCar.btnRentClick(Sender: TObject);
begin
  RentCar;
end;


end.


